File: 0648-replace-words/0648-replace-words.java
class Solution {
    static class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
    }
    TrieNode trie = new TrieNode();

    // Function to replace words
    private int replaceWord(String word) {
        TrieNode cur = this.trie;
        
        for (int i = 0; i < word.length(); i++) {
            char letter = word.charAt(i);
            if (cur.children.get(letter) == null)   break;
            
            cur = cur.children.get(letter);
            if (cur.children.containsKey('#'))  return i + 1;
        }
        return word.length();
    }
    
    public String replaceWords(List<String> roots, String sentence) {
        // Build the trie
        for (String root : roots) {
            TrieNode cur = this.trie;
            for (char letter : root.toCharArray()) {
                if (cur.children.containsKey('#')) break;
                cur = cur.children.computeIfAbsent(letter, k -> new TrieNode());
            }
            cur.children.put('#', null); // Mark the end of the word
        }

        // Split sentence into words and replace them
        StringBuilder result = new StringBuilder();
        for (String word : sentence.split("\\s+")) {
            if (result.length() > 0) result.append(" ");
            result.append(word.substring(0, replaceWord(word)));
        }

        return result.toString();
    }
}

File: 0648-replace-words/0648-replace-words.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        d = {"" : {}}
        for word in dictionary:
            cur = d
            for i in range(len(word)):
                if "end" in cur: break
                if word[i] not in cur:
                    cur[word[i]] = {}
                cur = cur[word[i]]
            cur["end"] = None
                    
        res = []
        for word in sentence.split(" "):
            added = False
            if word[0] in d:
                cur = d
                for i in range(len(word)):
                    if word[i] in cur:
                        cur = cur[word[i]]
                    else:
                        break
                    
                    if "end" in cur:
                        res.append(word[:i+1])
                        added = True
                        break
                        
            if not added:
                res.append(word)
        
        return " ".join(res)